File: client-sdk/python/examples/basic_usage.py
"""Basic usage example for the Full-Stack API Python SDK."""

from fullstack_api import Client
from fullstack_api.exceptions import AuthenticationError, ValidationError


def main():
    """Demonstrate basic SDK usage."""
    # Initialize client
    client = Client(base_url="http://localhost:8000")

    try:
        # Register a new user
        print("Registering new user...")
        user = client.auth.register(
            email="demo@example.com",
            username="demo_user",
            password="DemoPass123!",
            full_name="Demo User",
        )
        print(f"✓ Registered user: {user.username} ({user.email})")

    except ValidationError as e:
        print(f"✗ Registration failed: {e.message}")
        # User might already exist, try logging in

    try:
        # Login
        print("\nLogging in...")
        tokens = client.auth.login(
            username="demo@example.com",  # Can use email or username
            password="DemoPass123!",
        )
        print(f"✓ Logged in successfully")
        print(f"  Access token expires in: {tokens.expires_in} seconds")

        # Get current user
        print("\nFetching user profile...")
        me = client.users.get_current_user()
        print(f"✓ Current user: {me.full_name} ({me.email})")
        print(f"  Account created: {me.created_at}")
        print(f"  Email verified: {me.is_verified}")

        # Update profile
        print("\nUpdating profile...")
        updated = client.users.update_profile(
            full_name="Demo User Updated",
        )
        print(f"✓ Profile updated: {updated.full_name}")

        # Logout
        print("\nLogging out...")
        client.auth.logout()
        print("✓ Logged out successfully")

    except AuthenticationError as e:
        print(f"✗ Authentication failed: {e.message}")
    except Exception as e:
        print(f"✗ Error: {e}")


if __name__ == "__main__":
    main()

File: client-sdk/typescript/examples/basic-usage.ts
/**
 * Basic usage example for the Full-Stack API TypeScript SDK
 */

import { FullStackClient, AuthError, ValidationError } from '@fullstack/api-client';

async function main() {
  // Initialize client
  const client = new FullStackClient({
    baseUrl: 'http://localhost:8000',
  });

  try {
    // Register a new user
    console.log('Registering new user...');
    try {
      const user = await client.auth.register({
        email: 'demo@example.com',
        username: 'demo_user',
        password: 'DemoPass123!',
        fullName: 'Demo User',
      });
      console.log(`✓ Registered user: ${user.username} (${user.email})`);
    } catch (error) {
      if (error instanceof ValidationError) {
        console.log(`✗ Registration failed: ${error.message}`);
        // User might already exist, continue to login
      } else {
        throw error;
      }
    }

    // Login
    console.log('\nLogging in...');
    const tokens = await client.auth.login({
      username: 'demo@example.com', // Can use email or username
      password: 'DemoPass123!',
    });
    console.log('✓ Logged in successfully');
    console.log(`  Access token expires in: ${tokens.expiresIn} seconds`);

    // Get current user
    console.log('\nFetching user profile...');
    const me = await client.users.getCurrentUser();
    console.log(`✓ Current user: ${me.fullName} (${me.email})`);
    console.log(`  Account created: ${me.createdAt}`);
    console.log(`  Email verified: ${me.isVerified}`);

    // Update profile
    console.log('\nUpdating profile...');
    const updated = await client.users.updateProfile({
      fullName: 'Demo User Updated',
    });
    console.log(`✓ Profile updated: ${updated.fullName}`);

    // Add custom interceptor
    console.log('\nAdding request logger...');
    const interceptorId = client.addRequestInterceptor((config) => {
      console.log(`  → ${config.method?.toUpperCase()} ${config.url}`);
      return config;
    });

    // Make another request to see the interceptor in action
    await client.users.getCurrentUser();

    // Remove interceptor
    client.removeRequestInterceptor(interceptorId);

    // Logout
    console.log('\nLogging out...');
    await client.auth.logout();
    console.log('✓ Logged out successfully');

  } catch (error) {
    if (error instanceof AuthError) {
      console.error(`✗ Authentication failed: ${error.message}`);
    } else if (error instanceof ValidationError) {
      console.error(`✗ Validation failed: ${error.message}`);
      if (error.errors.length > 0) {
        console.error('  Errors:');
        error.errors.forEach(err => {
          console.error(`    - ${err.field}: ${err.message}`);
        });
      }
    } else {
      console.error(`✗ Error: ${error}`);
    }
  }
}

// Run the example
main().catch(console.error);
